deployment/datastore/memory_datastore.go

```
@@ -0,0 +1,107 @@
+	Sealer[DataStore]
+
+	BaseDataStore[MutableAddressRefStore]
+}
+
+// MemoryDataStore is a concrete implementation of the MutableDataStore interface.
+var _ MutableDataStore = &MemoryDataStore{}
+
+type MemoryDataStore struct {
+	AddressRefStore *MemoryAddressRefStore `json:"addressRefStore"`
+}
+
+// NewMemoryDataStore creates a new instance of MemoryDataStore.
+// NOTE: The instance returned is mutable and can be modified.
+func NewMemoryDataStore() *MemoryDataStore {
+	return &MemoryDataStore{
+		AddressRefStore: NewMemoryAddressRefStore(),
+	}
+}
+
+// Seal seals the MemoryDataStore, by returning a new instance of sealedMemoryDataStore.
+func (s *MemoryDataStore) Seal() DataStore {
+	return &sealedMemoryDataStore{AddressRefStore: s.AddressRefStore}
+}
+
+// Addresses returns the AddressRefStore of the MemoryDataStore.
+func (s *MemoryDataStore) Addresses() MutableAddressRefStore {
+	return s.AddressRefStore
+}
+
+// Merge merges the given mutable data store into the current MemoryDataStore.
+func (s *MemoryDataStore) Merge(other MutableDataStore) error {
+	addressRefs, err := other.Addresses().Fetch()
+	if err != nil {
+		return err
+	}
+
+	for _, addressRef := range addressRefs {
+		if err := s.AddressRefStore.AddOrUpdate(addressRef); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// SealedMemoryDataStore is a concrete implementation of the DataStore interface.
+// It represents a sealed data store that cannot be modified further.
+var _ DataStore = &sealedMemoryDataStore{}
+
+type sealedMemoryDataStore struct {
+	AddressRefStore *MemoryAddressRefStore `json:"addressRefStore"`
+}
+
+// Addresses returns the AddressRefStore of the sealedMemoryDataStore.
+// It implements the BaseDataStore interface.
+func (s *sealedMemoryDataStore) Addresses() AddressRefStore {
+	return s.AddressRefStore
+}
+
+// Merge merges the given data store into the current sealedMemoryDataStore.
+func (s *sealedMemoryDataStore) Merge(other DataStore) error {
```

i thought sealedMemoryDataStore is suppose to be read only, but then it supports Merge which modifies it?

---

Is that because both sealedMemoryDataStore and MemoryDataStore implement the same interface? In that case maybe we should return an error that they cannot update the sealedMemoryDataStore when merge is called 🤔

---

>Is that because both sealedMemoryDataStore and MemoryDataStore implement the same interface? In that case maybe we should return an error that they cannot update the sealedMemoryDataStore when merge is called 🤔
-----------   
> i thought sealedMemoryDataStore is suppose to be read only, but then it supports Merge which modifies it?

Yeah, this is not ideal. `sealedMemoryDataStore` is meant to be a read-only view of `MemoryDataStore`, so it really shouldn’t implement the `Merger` interface at all. This was originally based on some in initial assumptions about how `MemoryDataStore` would be integrated into `deployment.Environment`, `deployment.ChangesetOutput`, and `chainlink-deployments`. I’ve found a better approach now, so I’ll update this and remove the `Merger` implementation from `sealedMemoryDataStore`.